clients/ssh/git_upload_pack.go

```diff
@@ -0,0 +1,194 @@
+	}
+	s := "ssh://git@" + string(vcs.RepoHost) + ":22/" + vcs.FullName
+	u, err = url.Parse(s)
+	return
+}
+
+func connect(host, user string) (*ssh.Client, error) {
+
+	// try ssh-agent first
+	// then password auth
+	conn, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK"))
+	if err != nil {
+		return connectWithPasswd(host, user)
+	}
+	defer conn.Close()
+
+	agent := agent.NewClient(conn)
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{ssh.PublicKeysCallback(agent.Signers)},
+	}
+
+	client, err := ssh.Dial("tcp", host, sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial: %v", err)
+	}
+
+	return client, nil
+}
+
+// TODO: this must be done securely
+func connectWithPasswd(host, user string) (*ssh.Client, error) {
+	var pass string = "your password"
+
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{ssh.Password(pass)},
+	}
+
+	client, err := ssh.Dial("tcp", host, sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dia: %v", err)
+	}
+
+	return client, nil
+}
+
+func (s *GitUploadPackService) Info() (i *common.GitUploadPackInfo, err error) {
+	if !s.connected {
+		return nil, fmt.Errorf("not connected")
+	}
+
+	session, err := s.client.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open session: %v", err)
+	}
+	defer session.Close()
+
+	out, err := session.Output("git-upload-pack " + s.vcs.FullName + ".git")
+	if err != nil {
+		return nil, fmt.Errorf("ssh.session.Output: %v", err)
+	}
+
+	reader := bytes.NewReader(out)
+	dec := pktline.NewDecoder(reader)
+	return common.NewGitUploadPackInfo(dec)
+}
+
+func (s *GitUploadPackService) Disconnect() (err error) {
+	if !s.connected {
+		return fmt.Errorf("not connected")
+	}
+	if err = s.client.Close(); err != nil {
+		return err
+	}
+	s.client = nil
+	s.vcs = nil
+	return nil
+}
+
+// TODO: fetch should really reuse the info session instead of openning a new
+// one
+func (s *GitUploadPackService) Fetch(r *common.GitUploadPackRequest) (io.ReadCloser, error) {
+	if !s.connected {
+		return nil, fmt.Errorf("not connected")
+	}
+
+	session, err := s.client.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open session: %v", err)
+	}
+	defer session.Close()
+
+	si, err := session.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get ssh session stdin: %v", err)
+	}
+
+	so, err := session.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get ssh session stdout: %v", err)
+	}
+
+	go func() {
+		fmt.Fprintln(si, r.String())
+		si.Close()
+	}()
+
+	err = session.Start("git-upload-pack " + s.vcs.FullName + ".git")
+	if err != nil {
+		return nil, fmt.Errorf("ssh.session.Start: %v", err)
+	}
+	session.Wait()
+
+	// read until the header of the second answer
+	soBuf := bufio.NewReader(so)
+	token := "0000"
+	for {
+		line, err := soBuf.ReadString('\n')
+		if err == io.EOF {
+			return nil, fmt.Errorf("Bad answer format to git-upload-pack")
```

Comment:
Same as before: `var ErrGitUploadBadAnswer = errors.New("Bad answer format to git-upload-pack")`, error name is just a proposal.